### PR TITLE
Ecchen/improved rendering

### PR DESCRIPTION
**Pull request by CNOCycle**
1. new cache in rendering function
- Redesign the cache mechanism and cache the raw GRB image.

2. PixelPipeline performance improved
- performance improved by auto-rounding

### CODE REVIEW

**Review by grimmerk — COMMENTED** [review]
I notice that there is a piece of the code
```
m_frameCache.setMaxCost( 1 * 1024 * 1024 * 1024 );
```
should we enlarge it? @CNOCycle , @kswang1029

**Comment by grimmerk** [comment]
From Susan:  The redesign issue (9) is in a part of the code that I did not write so without a lot of work, I am not qualified to comment on it.  My worry with it would be that if you are not caching things like the zoom, pan, etc in the cache id, then you might pull out a cached image when you should not.  However, I assume you have also checked that something like that does not happen.

**Review by grimmerk — COMMENTED** [review]
This commit should not have problems, just curious that "auto-rounding is better than old way" is based on experiments/measurement or some document/theory?

File: carta/cpp/CartaLib/PixelPipeline/IPixelPipeline.h [hunk 1]
```diff
@@ -34,7 +34,7 @@ namespace PixelPipeline
 ///   - can have mulitple sub-steps
 ///   - configurable
 /// - stage 5: Drgb -> QRgb
-///   - simple conversion to 8 bit values using rounding
```
Comment (grimmerk) [inline carta/cpp/CartaLib/PixelPipeline/IPixelPipeline.h:37]: test review

**Comment by CNOCycle** [comment]
According to the C++ standard document:
`double round( double arg );`
SO, It still does the converting from double into int after calling round function.
std::round is more expensive than auto-rounding and the result may slightly different by auto-rounding if arg is very close to an integer value.
I think that it is safe in this case because we do not need high precision but high performance.

**Comment by grimmerk** [comment]
test

**Comment by pfederl** [comment]
From a quick look at the changes I have no issues with the proposed changes. Yes, the automatic rounding could perform better at the cost of losing some accuracy. The users would probably not notice the slight change in color.

**Comment by kswang1029** [comment]
For this fix, I confirm that the performance is enhanced by a factor of few (2-4) (checked with playing a movie using the animator)

**Comment by kswang1029** [comment]
Confirmed. The performance of zoom and pan is enhanced by this revised approach.

**Pull request merged**